### PR TITLE
OmegaConf

### PR DESCRIPTION
**Pull request by arturtoshev**
Replace simple yaml with OmegaConf and refactor the library.

### CODE REVIEW

File: lagrangebench/train/trainer.py [hunk 1]
```diff
@@ -40,18 +52,17 @@ def _mse(
     particle_type: jnp.ndarray,
     target: jnp.ndarray,
     model_fn: Callable,
-    loss_weight: LossConfig,
+    loss_weight: Dict[str, float],
```
Comment (gerkone) [inline lagrangebench/train/trainer.py:55]: Small thing but don't think this is the case. The `LossConfig` dataclass is passed

File: lagrangebench/case_setup/case.py [hunk 2]
```diff
@@ -63,13 +64,12 @@ def case_builder(
...
+    cfg_neighbors: DictConfig = defaults.neighbors,
+    isotropic_norm: bool = defaults.model.isotropic_norm,
+    noise_std: float = defaults.train.noise_std,
     external_force_fn: Optional[Callable] = None,
-    magnitude_features: bool = defaults.magnitude_features,
-    neighbor_list_backend: str = defaults.neighbor_list_backend,
-    neighbor_list_multiplier: float = defaults.neighbor_list_multiplier,
-    dtype: jnp.dtype = defaults.dtype,
+    magnitude_features: bool = defaults.model.magnitude_features,
+    dtype: jnp.dtype = defaults.main.dtype,
```
Comment (gerkone) [inline lagrangebench/case_setup/case.py:72]: why did you leave these as regular args?

File: main.py [hunk 3]
```diff
@@ -1,38 +1,58 @@
...
+        del cfgs[0]["extends"]
+
+        # go to parents configs until defaults.py is reached
+        if extends_path != "lagrangebench/defaults.py":
+            cfgs = [OmegaConf.load(extends_path)] + cfgs
+        else:
+            from lagrangebench.defaults import defaults
+
+            cfgs = [defaults] + cfgs
+            break
```
Comment (gerkone) [inline main.py:21]: Cool. Not intuitive at first since `defaults.py` is the only non-yaml file you can specify as extend. Do you insist in having these nested extends? This could be as simple as always merging the defaults to the passed config

File: lagrangebench/defaults.py [hunk 4]
```diff
@@ -1,70 +1,189 @@
...
+    cfg.main.seed = 0
+    # data type for preprocessing. One of "float32" or "float64"
+    cfg.main.dtype = "float64"
+    # gpu device. -1 for CPU. Should be specified before importing the library.
+    cfg.main.gpu = None
+    # XLA memory fraction to be preallocated. The JAX default is 0.75.
+    # Should be specified before importing the library.
+    cfg.main.xla_mem_fraction = None
+    # data directory
+    cfg.main.data_dir = None
```
Comment (gerkone) [inline lagrangebench/defaults.py:30]: What about not nesting these? Especially `main.config`, so one can just call `python main.py config=configs/...`

**Review by gerkone — COMMENTED** [review]
Will follow up on `LossConfig`, and some minor typing/library usability changes

File: lagrangebench/train/trainer.py [hunk 5]
```diff
@@ -89,129 +89,140 @@ def _update(
...
-    def _train(
-        step_max: int = defaults.step_max,
+        if wandb_run is None and self.cfg_logging.wandb:
+            wandb_run = wandb.init(
+                project=self.cfg_logging.wandb_project,
+                entity=self.cfg_logging.wandb_entity,
+                name=self.cfg_logging.run_name,
+                save_code=True,
+            )
+        self.wandb_run = wandb_run
```
Comment (gerkone) [inline lagrangebench/train/trainer.py:216]: What do we do in this case?

File: lagrangebench/case_setup/case.py (hunk 2 above)
Comment (arturtoshev) [reply lagrangebench/case_setup/case.py:72]: No particular reason. I don't want to pass the full configs if only 1-2 things are used.

File: lagrangebench/defaults.py (hunk 4 above)
Comment (arturtoshev) [reply lagrangebench/defaults.py:30]: I chose the name "main" as some of these things are only used in the main.py, i.e. main.config and main.xla_mem_fraction. But we can move all of them out - I indeed didn't see the "main" nesting in other projects.

File: lagrangebench/defaults.py [hunk 6]
```diff
@@ -1,70 +1,189 @@
...
+    assert cfg.main.mode in ["train", "infer", "all"]
+    assert cfg.main.dtype in ["float32", "float64"]
+    assert cfg.main.data_dir is not None, "main.data_dir must be specified."
+
+    assert cfg.eval.train.n_trajs >= -1
+    assert cfg.eval.infer.n_trajs >= -1
+    assert set(cfg.eval.train.metrics).issubset(["mse", "e_kin", "sinkhorn"])
+    assert set(cfg.eval.infer.metrics).issubset(["mse", "e_kin", "sinkhorn"])
+    assert cfg.eval.train.out_type in ["none", "vtk", "pkl"]
+    assert cfg.eval.infer.out_type in ["none", "vtk", "pkl"]
```
[the remark this answers is not in the document]
Comment (arturtoshev) [reply lagrangebench/defaults.py:189]: Currently, the check_cfg is just a proof of concept. Feel free to extend it. But I wouldn't put things that might limit the functionality of sth.

File: main.py (hunk 3 above)
Comment (arturtoshev) [reply main.py:21]: I do insist on the option of having nested configs, and I'm still not too happy about not having the base.yaml for every dataset. 

Do you think that always having defaults.py as the lowest priority overwrite is a good idea? Then if we use one level of configs, there would be no example of how to even use "extends: ".

File: tests/runner_test.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                    "ckp_dir": "/tmp/ckp",
+                },
+            }
+        )
+        # overwrite defaults with user-defined config
+        self.cfg = OmegaConf.merge(defaults, self.cfg)
+
+    def test_runner(self):
+        out = train_or_infer(self.cfg)
+        self.assertEqual(out, 0)
```
[the remark this answers is not in the document]
Comment (arturtoshev) [reply tests/runner_test.py:57]: Exactly, I do only check whether it runs through. It does go over an actual training and inference but doesn't log to wand and writes the checkpoint to `/tmp`.

File: lagrangebench/defaults.py (hunk 4 above)
Comment (arturtoshev) [reply lagrangebench/defaults.py:30]: One advantage of cfg.main.* is easier searching for stuff through the code, but this is a minor point

File: lagrangebench/train/trainer.py (hunk 5 above)
Comment (arturtoshev) [reply lagrangebench/train/trainer.py:216]: I'll move that to .train

File: lagrangebench/defaults.py (hunk 4 above)
Comment (arturtoshev) [reply lagrangebench/defaults.py:30]: ok, I'll remove the .main. hierarchy level

File: main.py (hunk 3 above)
Comment (arturtoshev) [reply main.py:21]: ok, I'll use a special string to denote the lagrangebench default configs

File: lagrangebench/defaults.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arturtoshev) [reply lagrangebench/defaults.py:189]: I'll go over the defaults and see what I can add to the checks.

File: lagrangebench/train/trainer.py (hunk 1 above)
Comment (arturtoshev) [reply lagrangebench/train/trainer.py:43]: Is this still relevant?

File: lagrangebench/__init__.py [hunk 8]
```diff
@@ -21,7 +20,9 @@
...
     "LDC3D",
     "DAM2D",
-    "PushforwardConfig",
 ]
 
-__version__ = "0.0.1"
+import toml
+
+pyproject = toml.load("pyproject.toml")
+__version__ = pyproject["tool"]["poetry"]["version"]
```
Comment (gerkone) [inline lagrangebench/__init__.py:28]: Will this work in the library version?

**Review by gerkone — APPROVED** [review]
